core/src/main/java/org/apache/iceberg/RESTFileScanTaskParser.java

```diff
@@ -0,0 +1,98 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.ExpressionParser;
+import org.apache.iceberg.expressions.ResidualEvaluator;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.JsonUtil;
+
+public class RESTFileScanTaskParser {
+  private static final String DATA_FILE = "data-file";
+  private static final String DELETE_FILE_REFERENCES = "delete-file-references";
+  private static final String RESIDUAL_FILTER = "residual-filter";
+
+  private RESTFileScanTaskParser() {}
+
+  public static void toJson(
+      FileScanTask fileScanTask,
+      Set<Integer> deleteFileReferences,
+      PartitionSpec partitionSpec,
+      JsonGenerator generator)
+      throws IOException {
+    Preconditions.checkArgument(fileScanTask != null, "Invalid file scan task: null");
+    Preconditions.checkArgument(generator != null, "Invalid JSON generator: null");
+
+    generator.writeStartObject();
+    generator.writeFieldName(DATA_FILE);
+    ContentFileParser.toJson(fileScanTask.file(), partitionSpec, generator);
+    if (deleteFileReferences != null) {
+      JsonUtil.writeIntegerArray(DELETE_FILE_REFERENCES, deleteFileReferences, generator);
+    }
+
+    if (fileScanTask.residual() != null) {
+      generator.writeFieldName(RESIDUAL_FILTER);
+      ExpressionParser.toJson(fileScanTask.residual(), generator);
+    }
+
+    generator.writeEndObject();
+  }
+
+  public static FileScanTask fromJson(
+      JsonNode jsonNode,
+      List<DeleteFile> allDeleteFiles,
+      Map<Integer, PartitionSpec> specsById,
+      boolean isCaseSensitive) {
+    Preconditions.checkArgument(jsonNode != null, "Invalid JSON node for file scan task: null");
+    Preconditions.checkArgument(
+        jsonNode.isObject(), "Invalid JSON node for file scan task: non-object (%s)", jsonNode);
+
+    DataFile dataFile =
+        (DataFile) ContentFileParser.fromJson(JsonUtil.get(DATA_FILE, jsonNode), specsById);
+    int specId = dataFile.specId();
+
+    DeleteFile[] deleteFiles = null;
+    if (jsonNode.has(DELETE_FILE_REFERENCES)) {
+      List<Integer> indices = JsonUtil.getIntegerList(DELETE_FILE_REFERENCES, jsonNode);
+      deleteFiles =
+          indices.stream()
+              .map(index -> (GenericDeleteFile) allDeleteFiles.get(index))
```

Comment:
I think we should be a bit defensive here in case the response has `DELETE_FILE_REFERENCES` but has no deletes. Right now, it looks like we return null https://github.com/apache/iceberg/blob/580097c8449bbada1357192680d16363a501fc8c/core/src/main/java/org/apache/iceberg/TableScanResponseParser.java#L55 when there are no top level deletes.

Could we add a Preconditions here that if there are delete file references that allDeleteFiles != null?

Reply:
Well actually, I think it's really that there should be at least 1 delete file in `allDeleteFiles`. In that case could we change that implementation I linked to return an empty list. Then the preconditions check can just be `!allDeleteFiles.isEmpty()`

Reply:
For COW table we might not have delete files at all, in that case my expectation what we should not have deleteFileReferences and neither all deletes but if we have deleteFileReference but `allDeleteFiles` are empty then its an illegal state, IMHO we should be fine with throwing Array out of bound, sending an empty list instead of null would be suff, wdyt ?

Reply:
Yeah it's certainly an illegal state but since it's a public API, my main point was that I think it's good to not make any assumptions about caller state and do the validation so that there's a clearer error message. Looks like this was fixed in the recent update

Reply:
Oh nvm, looks like this wasn't fixed. Let me know if my argument around this being a public API having validation with a clear error message doesn't make sense. I don't think we want public APIs to fail with an odd array index out of bounds since we don't want those APIs to make assumptions on the context of the caller

Reply:
i think then it would be best to check the max in reference array is less than allDeleteFilesSize ? added the check.